### DISCUSSION TITLE
Multiplayer server connection problem

### DISCUSSION THREAD

**Discussion by denglihong2007** [perpetualKid/FreeTrainSimulator#109] [Q&A] [answered]
First of all, I'm sure I have configured my server correctly, but I use the OR from China, which can often run normally with this server. However, it sometimes has the following problems. 1. When the second player uses the server, the first player is prompted that there is only one player in the game, and the second player is prompted that there are two players in the game, and the actions of the second player will not be presented at the first player. 2. When a single person uses the server, sometimes the scheduling permission is lost while playing. Even in multiplayer games, another ordinary player will be prompted to become a dispatcher, and the operation of one dispatcher cannot be synchronized to other players. 3. Although this server is unoccupied, the first player who logs in does not have scheduling permission, nor do other players. Even each player is prompted that only one player is playing. In addition, can we change the dispatcher on the server side? Thank you.

**Comment by perpetualKid** [comment, marked as answer]
it's difficult to see what the problem might be. 
While the multiplayer server does not have a logfile itself, one can redirect the console output and capture in logfile which you may want to attach so we can read browse through.
However, I know there are some issues in the multiplayer code stack, and they are difficult to fix, since the original server is not open source and not completely understood, but still widely in use with OpenRails. So any fixes would immediately break compatiblity with OpenRails, which we are hesitant so far. 
Longer term though there should be an update to the multiplayer part, which will (should) also allow to run the dispatcher separately, but this work has not been started yet due to resource constraints.

**Reply by denglihong2007** [reply]
Thanks for your help.

**Reply by perpetualKid** [reply]
> Thanks for your help.

any contribution is welcome, as much as any bug report and discussion about feature requests!

**Reply by denglihong2007** [reply]
> > Thanks for your help.
> 
> any contribution is welcome, as much as 
any bug report and discussion about feature requests!

For a small request, I decompiled OpenRails For China and extracted three CS files. They seem to be related to the standalone server program. I wonder if I can obtain better compatibility by using these files to make standalone server programs. How can I make standalone server programs? The files is attached.By the way, some codes may look strange and chaotic, and they are meaningless.[Files.zip](https://github.com/perpetualKid/ORTS-MG/files/9438158/Files.zip)

**Reply by perpetualKid** [reply]
thank you! The code looks mostly same as the client side code in OpenRails, so alternatively to looking at the decompiled code you could also use the source from here https://github.com/openrails/openrails/tree/master/Source/Orts.Simulation/MultiPlayer
The challenge with server program is that there is no documentation in source or elsewhere about the protocol (the messages which are exchange between server and client/s), and it behaves slightly different having a standalone server vs one of the players being the server. 
Either one needs to monitor and trace active multiplayer sessions and log all the possible cases to understand what is being sent back and forth (which we did to come up with our version of standalone server, but not call cases might be covered), or (longer term) develop and document a new/updated version of the protocol and implement.
Since there seem very few users only using mulitplayer at all, there is a low priority right now to focus on this, but happy to help if someone wants to contribute or look into this.
